src/iterators/iterator.md

```diff
@@ -2,49 +2,63 @@
 minutes: 5
 ---
 
-# `Iterator`
+# `Iterator` Trait
 
-The [`Iterator`][1] trait supports iterating over values in a collection. It
-requires a `next` method and provides lots of methods. Many standard library
-types implement `Iterator`, and you can implement it yourself, too:
+The [`Iterator`][1] trait defines how an object can be used to produce a
+sequence of values. For example, if we wanted to create an iterator that can
+produce the elements of a slice it might look something like this:
 
 ```rust,editable
-struct Fibonacci {
-    curr: u32,
-    next: u32,
+struct SliceIter<'a> {
+    slice: &'a [i32],
+    i: usize,
 }
 
-impl Iterator for Fibonacci {
-    type Item = u32;
+impl<'a> Iterator for SliceIter<'a> {
+    type Item = &'a i32;
 
     fn next(&mut self) -> Option<Self::Item> {
-        let new_next = self.curr + self.next;
-        self.curr = self.next;
-        self.next = new_next;
-        Some(self.curr)
+        if self.i == self.slice.len() {
+            None
+        } else {
+            let next = &self.slice[self.i];
+            self.i += 1;
+            Some(next)
+        }
     }
 }
 
 fn main() {
-    let fib = Fibonacci { curr: 0, next: 1 };
-    for (i, n) in fib.enumerate().take(5) {
-        println!("fib({i}): {n}");
+    let slice = [2, 4, 6, 8].as_slice();
+    let iter = SliceIter { slice, i: 0 };
+    for elem in iter {
+        println!("elem: {elem}");
     }
 }
 ```
 
 <details>
 
-- The `Iterator` trait implements many common functional programming operations
-  over collections (e.g. `map`, `filter`, `reduce`, etc). This is the trait
-  where you can find all the documentation about them. In Rust these functions
-  should produce the code as efficient as equivalent imperative implementations.
+- The `SliceIter` example implements the same logic as the C-style `for` loop
+  demonstrated on the last slide.
 
-- `IntoIterator` is the trait that makes for loops work. It is implemented by
-  collection types such as `Vec<T>` and references to them such as `&Vec<T>` and
-  `&[T]`. Ranges also implement it. This is why you can iterate over a vector
-  with `for i in some_vec { .. }` but `some_vec.next()` doesn't exist.
+- Point out to the students that iterators are lazy: Creating the iterator just
+  initializes the struct but does not otherwise do any work. No work happens
+  until the `next` method is called.
```

Comment:
Worth mentioning here that iterators need not be finite.

Reply:
Good point, added a speaker note about it.